File: 2003/VB101SamplesAll/VB101SamplesWinForms/UsingMaskedTextBox/UsingMaskedTextBox/LatLong.vb
Imports System.Windows.Forms.Design
Imports System.Text


#Region "Latitude type"

' Implement custom types for Latitude.
Public Class Latitude

    Public Enum DirectionEnum
        North
        South
    End Enum

    Private v As Single
    Private d As DirectionEnum

    Public Property Direction() As DirectionEnum
        Get
            Return d
        End Get
        Set(ByVal value As DirectionEnum)
            d = value
        End Set
    End Property

    Public Property Value() As Single
        Get
            Return v
        End Get
        Set(ByVal value As Single)
            If (0 <= value And value <= 90) Then
                v = value
            Else
                Throw New Exception("Invalid Latitude value.")
            End If
        End Set
    End Property

    Public Sub New(ByVal value As Single, ByVal direction As DirectionEnum)

        Me.Value = value
        Me.Direction = direction
    End Sub

    ' Constructor: "97.3 N", "123 S"
    Public Sub New(ByVal latitude As String)

        Dim parts As String() = latitude.Split(New String() {" "}, StringSplitOptions.RemoveEmptyEntries)
        Me.Value = Single.Parse(parts(0))

        Select Case parts(1).ToUpper()
            Case "N"
                Me.Direction = DirectionEnum.North
                Exit Select
            Case "S"
                Me.Direction = DirectionEnum.South
                Exit Select
            Case Else
                Throw New ApplicationException("Invalid latitude. You must specify N or S.")
        End Select
    End Sub

    Public Overrides Function ToString() As String
        Dim directionString As String
        If Me.Direction = DirectionEnum.North Then
            directionString = "N"
        Else
            directionString = "S"
        End If

        Return Me.Value.ToString() + " " + directionString
    End Function

    ' A public static Parse method is required for MaskedTextBox 
    ' custom type validation
    Public Shared Function Parse(ByVal latitude As String) As Latitude
        Return New Latitude(latitude)
    End Function

End Class

#End Region

#Region "LatitudeMaskDescriptor"

' Implement a MaskDescriptor for the Latitude type.  
' The InputMask dialog will display all types for which
' a MaskDescriptor is found.
Public Class LatitudeMaskDescriptor
    Inherits MaskDescriptor

    Public Overrides ReadOnly Property Mask() As String
        Get
            Return "00 >L"
        End Get
    End Property

    Public Overrides ReadOnly Property Name() As String
        Get
            Return "Latitude"
        End Get
    End Property

    Public Overrides ReadOnly Property Sample() As String
        Get
            Return "58 N"
        End Get
    End Property

    Public Overrides ReadOnly Property ValidatingType() As Type
        Get
            Return GetType(Latitude)
        End Get
    End Property

End Class

#End Region

File: 2003/CS101SamplesAll/CS101SamplesWinForms/UsingMaskedTextBox/MaskedTextBox/LatLong.cs
using System;
using System.Windows.Forms.Design;
using System.Text;

namespace UsingMaskedTextBox
{

    //  Implement custom types for Latitude.
    public class Latitude
    {
        private float v;
        private DirectionEnum d;

        public Latitude(float value, DirectionEnum direction)
        {
            this.Value = value;
            this.Direction = direction;
        }

        //  Constructor: "97.3 N", "123 S"
        public Latitude(string latitude)
        {
            char[] seperator = new char[2];
            seperator[0] = ' ';
            string[] parts = latitude.Split(seperator, StringSplitOptions.RemoveEmptyEntries);

            this.Value = float.Parse(parts[0]);
            switch (parts[1].ToUpper())
            {
                case "N":
                    this.Direction = DirectionEnum.North;
                    break;
                case "S":
                    this.Direction = DirectionEnum.South;
                    break;
                default:
                    throw new ApplicationException("Invalid latitude. You must specify N or S.");
            }
        }

        public DirectionEnum Direction
        {
            get
            {
                return d;
            }
            set
            {
                d = value;
            }
        }

        public float Value
        {
            get
            {
                return v;
            }
            set
            {
                if (((0 <= value) && (value <= 90)))
                {
                    v = value;
                }
                else
                {
                    throw new Exception("Invalid Latitude value.");
                }
            }
        }

        public override string ToString()
        {
            string directionString;
            if ((this.Direction == DirectionEnum.North))
            {
                directionString = "N";
            }
            else
            {
                directionString = "S";
            }
            return (this.Value.ToString() + (" " + directionString));
        }

        //  A public static Parse method is required for MaskedTextBox 
        //  custom type validation
        public static Latitude Parse(string latitude)
        {
            return new Latitude(latitude);
        }

        public enum DirectionEnum
        {

            North,

            South,
        }
    }
    //  Implement a MaskDescriptor for the Latitude type.  
    //  The InputMask dialog will display all types for which
    //  a MaskDescriptor is found.
    public class LatitudeMaskDescriptor : MaskDescriptor
    {

        public override string Mask
        {
            get
            {
                return "00 >L";
            }
        }

        public override string Name
        {
            get
            {
                return "Latitude";
            }
        }

        public override string Sample
        {
            get
            {
                return "58 N";
            }
        }

        public override Type ValidatingType
        {
            get
            {
                return typeof(Latitude);
            }
        }
    }
}